[change] 예외처리 기능 추가

LEEDONGH00N opened the pull request:
# 이슈
- #1 

# 구현 기능

ErrorCode 관련 기능 추가
- ErrorCode: 에러 코드 정의 클래스 추가
- ErrorCodeDto: 클라이언트와의 에러 코드 전송을 위한 DTO 클래스 추가
- ErrorDisplayType: 에러 표시 유형을 정의하는 열거형 클래스 추가

커스텀 예외 클래스 BaseException 추가
- BaseException 클래스 구현
- ErrorCode를 기반으로 예외를 관리하도록 설계
- 코드와 커스텀 메시지를 포함하여 예외 정보를 통합적으로 처리

전역 예외 처리 기능 추가
- GlobalExceptionHandler 클래스 구현
- 예외 발생 시 공통된 형태의 API을 제공하도록 설계
- 다양한 예외를 구분하여 처리하고, 적절한 상태 코드 및 에러 메시지를 반환

ErrorResponse 클래스 추가
- 에러 응답을 처리하기 위한 DTO 클래스 ErrorResponse 구현
- 에러 정보를 클라이언트에게 일관된 형식으로 전달

# 작업 시간
2h

LEEDONGH00N committed:
[feat] 예외처리 기능 추가

[feat] ErrorCode 관련 기 능 추가
- ErrorCode: 에러 코드 정의 클래스 추가
- ErrorCodeDto: 클라이언트와의 에러 코드 전송을 위한 DTO 클래스 추가
- ErrorDisplayType: 에러 표시 유형을 정의하는 열거형 클래스 추가

[feat] 커스텀 예외 클래스 BaseException 추가
- BaseException 클래스 구현
- ErrorCode를 기반으로 예외를 관리하도록 설계
- 코드와 커스텀 메시지를 포함하여 예외 정보를 통합적으로 처리

[feat] 전역 예외 처리 기능 추가
- GlobalExceptionHandler 클래스 구현
- 예외 발생 시 공통된 형태의 API을 제공하도록 설계
- 다양한 예외를 구분하여 처리하고, 적절한 상태 코드 및 에러 메시지를 반환

[feat] ErrorResponse 클래스 추가
- 에러 응답을 처리하기 위한 DTO 클래스 ErrorResponse 구현
- 에러 응답을 처리하기 위한 DTO 클래스 ErrorResponse 구현
- 에러 정보를 클라이언트에게 일관된 형식으로 전달

HeeChanN commented:
ErrorCodeDto와 ErrorResponse 클래스는 구분 지은 이유가 있을까요? 
두 클래스의 필드를 살펴보면 하나의 클래스를 이용해도 될 거 같습니다.

HeeChanN reviewed (changes requested):
ErrorCodeDto와 ErrorResponse 클래스는 구분 지은 이유가 있을까요?
두 클래스의 필드를 살펴보면 하나의 클래스를 이용해도 될 거 같습니다.

h26n commented:
> ErrorCodeDto와 ErrorResponse 클래스는 구분 지은 이유가 있을까요? 두 클래스의 필드를 살펴보면 하나의 클래스를 이용해도 될 거 같습니다.

기존에 생각한 예외 처리 과정을 간략히 설명하자면
BaseException이 발생하면 GlobalException은 해당 BaseException에서 ErrorCode를 토대로 생성한 ErrorCodeDto를 통해 ErrorResponse를 생성해 반환 받습니다.

구분 지었던 이유는 ErrorCode의 필드 중 Response로 전송할 필드만 추려서 Dto로 생성한 다음 이 Dto를 바탕으로 BaseException에서 ErrorResponse를 생성할 계획이었습니다. 하지만 저희 컨벤션 내용 중 오류만 공통 Response 형식을 사용하기로 했고, 이 Response 형식과 ErrorCodeDto와 똑같다는 것을 알았습니다. 

말씀하신 내용을 바탕으로 수정하겠습니다.

LEEDONGH00N committed:
[feat] ErrorCodeDto 삭제 및 예외 처리 로직 단순화
ErrorCodeDto를 삭제해 ErrorResponse와의 중복 제거
ErrorCodeDto 반환 메서드 -> ErrorResponse 반환으로 변경

LEEDONGH00N committed:
[change] 예외 처리 로직 수정
BaseException이 ErrorResponse를 의존하지 않게 수정
